src/hotspot/share/runtime/thread.cpp

```diff
@@ -478,12 +478,13 @@ Thread::~Thread() {
 
 #ifdef ASSERT
 // A JavaThread is considered "dangling" if it is not the current
-// thread, has been added the Threads list, the system is not at a
-// safepoint and the Thread is not "protected".
+// thread, his not handshaking with current thread, as been added the Threads
+// list, the system is not at a safepoint and the Thread is not "protected".
 //
 void Thread::check_for_dangling_thread_pointer(Thread *thread) {
-  assert(!thread->is_Java_thread() || Thread::current() == thread ||
-         !thread->as_Java_thread()->on_thread_list() ||
+  assert(!thread->is_Java_thread() ||
+         ((JavaThread *) thread)->is_handshake_safe_for(Thread::current()) ||
+         !((JavaThread *) thread)->on_thread_list() ||
```

Comment:
Should use "thread->as_Java_thread()" instead of the cast here (2 places).

Reply:
Fixed